dev-db/postgresql: Change systemd unit type to notify

cjmayo opened the pull request:
Add systemd USE flag to control required configure option and
dependency.

Package-Manager: Portage-2.3.5, Repoman-2.3.2

---

PostgreSQL 9.6 added configure option --with-systemd to enable calling sd_notify() at server start and stop
https://www.postgresql.org/docs/9.6/static/release-9-6.html#AEN131350

Changes to systemd unit based on:
https://www.postgresql.org/docs/9.6/static/server-start.html

Replaces use of pg_ctl, default logs now sent to journal instead of postmaster.log.

```
--- postgresql.service
+++ postgresql.service-9.6
@@ -17,7 +17,7 @@
 After=network.target
 
 [Service]
-Type=forking
+Type=notify
 
 User=postgres
 Group=postgres
@@ -37,9 +37,10 @@
 # StandardOutput=syslog
 
 ExecStartPre=/usr/bin/postgresql-@SLOT@-check-db-dir
-ExecStart=/usr/@LIBDIR@/postgresql-@SLOT@/bin/pg_ctl start -D ${DATA_DIR} -s -l ${DATA_DIR}/postmaster.log -o "-p ${PGPORT} -D ${PGDATA} --data-directory=${DATA_DIR}" -w -t 300
-ExecStop=/usr/@LIBDIR@/postgresql-@SLOT@/bin/pg_ctl stop -D ${DATA_DIR} -s -m fast
-ExecReload=/usr/@LIBDIR@/postgresql-@SLOT@/bin/pg_ctl reload -D ${DATA_DIR} -s
+ExecStart=/usr/@LIBDIR@/postgresql-@SLOT@/bin/postgres -p ${PGPORT} -D ${DATA_DIR}
+ExecReload=/bin/kill -HUP $MAINPID
+KillMode=mixed
+KillSignal=SIGINT
```

```
--- postgresql-9.6.2-r1.ebuild
+++ postgresql-9.6.2-r2.ebuild
@@ -24,7 +24,7 @@
 LINGUAS="af cs de en es fa fr hr hu it ko nb pl pt_BR ro ru sk sl sv tr
 		 zh_CN zh_TW"
 IUSE="doc kerberos kernel_linux ldap libressl nls pam perl -pg_legacytimestamp python
-	  +readline selinux +server ssl static-libs tcl threads uuid xml zlib"
+	  +readline selinux +server systemd ssl static-libs tcl threads uuid xml zlib"
 
 for lingua in ${LINGUAS}; do
 	IUSE+=" linguas_${lingua}"
@@ -55,6 +55,7 @@
 	!libressl? ( >=dev-libs/openssl-0.9.6-r1:0= )
 	libressl? ( dev-libs/libressl:= )
 )
+systemd? ( sys-apps/systemd )
 tcl? ( >=dev-lang/tcl-8:0= )
 xml? ( dev-libs/libxml2 dev-libs/libxslt )
 zlib? ( sys-libs/zlib )
@@ -175,6 +176,7 @@
 		$(use_with python) \
 		$(use_with readline) \
 		$(use_with ssl openssl) \
+		$(use_with systemd) \
 		$(use_with tcl) \
 		${uuid_config} \
 		$(use_with xml libxml) \
@@ -251,7 +253,7 @@
 			"${FILESDIR}/${PN}.init-9.3-r1" | newinitd - ${PN}-${SLOT}
 
 		sed -e "s|@SLOT@|${SLOT}|g" -e "s|@LIBDIR@|$(get_libdir)|g" \
-			"${FILESDIR}/${PN}.service" | \
+			"${FILESDIR}/${PN}.service-9.6" | \
 			systemd_newunit - ${PN}-${SLOT}.service
```

mgorny commented on dev-db/postgresql/postgresql-9.6.2-r2.ebuild:
```diff
@@ -0,0 +1,477 @@
+		sed -e "s|@SLOT@|${SLOT}|g" -e "s|@LIBDIR@|$(get_libdir)|g" \
+			"${FILESDIR}/${PN}.init-9.3-r1" | newinitd - ${PN}-${SLOT}
+
+		sed -e "s|@SLOT@|${SLOT}|g" -e "s|@LIBDIR@|$(get_libdir)|g" \
```
Unless I'm missing something, your new unit file works only when USE=systemd is enabled. I'd say you either should make this conditional to `use systemd` then, or install the old unit if that flag is disabled.

titanofold replied:
I agree. Either the systemd files or the init files should be installed.

floppym replied:
> I agree. Either the systemd files or the init files should be installed.

I think you missed the point. The init scripts should continue to be installed unconditionally.

There are two options for the systemd units:
1. Hard-code `Type=notify` and the related changes proposed in this PR, and only install them when USE=systemd. This is because `Type=notifiy` requires libsystemd to work.
2. Conditionally set `Type=forking` if USE=systemd is disabled. This would be more tricky to support reliably, and seems a bit pointless.

mgorny commented:
@gentoo/systemd

titanofold commented on dev-db/postgresql/postgresql-9.6.2-r2.ebuild:
```diff
@@ -0,0 +1,477 @@
+	!libressl? ( >=dev-libs/openssl-0.9.6-r1:0= )
+	libressl? ( dev-libs/libressl:= )
+)
+systemd? ( sys-apps/systemd )
```
This should only matter if we're building the server. Wrap this with a `server? ( )`.

titanofold commented on dev-db/postgresql/postgresql-9.6.2-r2.ebuild:
```diff
@@ -0,0 +1,477 @@
+		einfo "Or move the configuration files back:"
+		einfo "mv ${PGDATA}*.conf ${DATA_DIR}"
+	else
+		einfo "You should use the '${EROOT%/}/etc/init.d/postgresql-${SLOT}' script to run PostgreSQL"
```
We should have instructions for starting the server with systemd here.

titanofold commented on dev-db/postgresql/postgresql-9.6.2-r2.ebuild:
```diff
@@ -0,0 +1,477 @@
+		$(use_with python) \
+		$(use_with readline) \
+		$(use_with ssl openssl) \
+		$(use_with systemd) \
```
If it's a non-server build, we shouldn't let this be enabled.

titanofold commented:
There's also the matter of unix_socket_directories being removed from postgresql.conf. We'll need a way for that to be handled somehow with the systemd bits.

eliasp commented:
> There's also the matter of unix_socket_directories being removed from postgresql.conf. We'll need a way for that to be handled somehow with the systemd bits.

This should be already taken care of with:
```INI
# Make sure the required runtimedir is present                                                                                                                                     
RuntimeDirectory=postgresql                                                                                                                                                        
RuntimeDirectoryMode=1775
```

titanofold commented:
That just creates **a** directory, which is fine since it's the default. However, 9.3 on up support multiple directories, and the unix_socket_directories setting has been stripped from /etc/postgresql-$slot/postgresql.conf. We need a way that's equivalent to the initscripts to emulate the functionality of creating the directories and passing the unix_socket_directories via the command line outside of postgresql.conf.

cjmayo commented:
Updated to 9.6.3 and added the live ebuild. I believe I've addressed the comments, also changed the commit message and einfo for postmaster.log no longer being used.

As for unix_socket_directories: isn't it OK that the user can create a drop-in overriding ExecStart and RuntimeDirectory (or a file in tmpfiles.d)? Anyway this isn't a new issue caused by changing the unit to notify.

floppym commented:
> As for unix_socket_directories: isn't it OK that the user can create a drop-in overriding ExecStart and RuntimeDirectory (or a file in tmpfiles.d)? Anyway this isn't a new issue caused by changing the unit to notify.

Agreed on both points.

cjmayo commented:
Updated to always install the init script.

mgorny commented on dev-db/postgresql/files/postgresql.service-9.6:
```diff
@@ -0,0 +1,56 @@
+# containing:
+#       [Service]
+#       Environment=PGPORT=5433
+# This will override the setting appearing below.
```
FYI, you can use `systemd_install_serviced` to install a boilerplate .d file with some useful stuff commented out.
